[team6 pair1] qna 4단계 구현 완료했습니다.

mindock opened the pull request.

Commit:
qna 4단계 구현 완료했습니다.

javajigi commented on src/main/java/codesquad/domain/Post.java:
```diff
@@ -0,0 +1,74 @@
+
+import javax.persistence.*;
+
+@MappedSuperclass
```
오호.
가르쳐주지 않아도 알아서 적용하는군요. 👍

javajigi commented on src/main/java/codesquad/domain/Question.java:
```diff
@@ -1,63 +1,71 @@
-    @ManyToOne
-    @JoinColumn(foreignKey = @ForeignKey(name = "fk_question_writer"))
-    private User writer;
+public class Question extends Post {
```
많은 중복 코드가 제거되니 깔끔하고 좋네요. 👍

javajigi commented on src/main/java/codesquad/domain/Question.java:
```diff
@@ -1,63 +1,71 @@
+    public void removeAnswer(Long answerId, User user) {
+        Answer answer = findAnswerById(answerId);
+        answer.validWriter(user);
+        answer.setDeleted(true);
```
삭제 상태를 파악하고 상태 변경 주제를 answer가 담당하도록 구현해 본다.
answer.delete(user);

javajigi commented on src/main/java/codesquad/web/QuestionController.java:
```diff
@@ -48,11 +53,13 @@ public String update(@PathVariable Long id, Question question, HttpSession sessi
-        questionRepository.delete(questionOrigin);
+        questionOrigin.validWriter(user);
+        questionOrigin.isPossibleDelete(user);
+        questionOrigin.setDeleted(true);
```
질문 삭제에 대한 주체를 question이 담당하도록 구현해 본다.
question.delete(loginUser);

javajigi reviewed (changes requested):
삭제 기능을 Question, Answer가 담당하도록 하면 더 많은 단위 테스트를 구현하는 것이 가능할 겁니다. 좀 더 객체지향적인 개발 연습이니 한번 도전해 보면 좋겠네요. 👍

Commit:
QnA step4 피드백 적용했습니다.

1. SessionUtil에서 로그인되지 않은 상태에서는 NullPointerException이 아닌 NoSessionedUserException을 발생시킵니다.
2. User의 list에서 수정불가하게 변경했습니다.
3. 삭제상태를 파악하고 상태를 변경하는 것을 Question, Answer에서 담당하도록 변경했습니다.
   단, 부모 클래스인 Post에서 공통적인 부분을 만들어 중복을 제거했습니다.
